docs/advanced/configuration.md
# Configuration

`superduperdb/misc/config.py` contains a complete definition using `pydantic` of the default
configuration. This may be overridden using a `configs.json` file containing in the
working directory.

There are three ways to set a config variable

* put just the values you want to change in a file `configs.json` at the room of `superduperdb-stealth`
* set an environment variable with the value
* set it in code

For example, these three forms are identical:

* Storing `{"remote": True, "dask": {"ip": "1.1.1.1"}}` in `configs.json`
* Setting environment variables `SUPERDUPERDB_REMOTE=true` and
  `SUPERDUPERDB_DASK_IP=1.1.1.1`
* In Python, `CFG.remote = True; CFG.dask.ip = '1.1.1.1'`

Here is what the default config object looks like in `JSON` format:

```json
{
    "apis": {
        "providers": {},
        "retry": {
            "wait_multiplier": 1.0,
            "wait_min": 4.0,
            "wait_max": 10.0,
            "stop_after_attempt": 2
        }
    },
    "dask": {
        "port": 8786,
        "password": "",
        "username": "",
        "ip": "localhost",
        "serializers": [],
        "deserializers": []
    },
    "logging": {
        "kwargs": {},
        "level": "INFO",
        "type": "STDERR"
    },
    "data_layers": {
        "artifact": {
            "cls": "mongodb",
            "connection": "pymongo",
            "kwargs": {
                "port": 27017,
                "password": "",
                "username": "",
                "host": "localhost"
            },
            "name": "_filesystem:documents"
        },
        "data_backend": {
            "cls": "mongodb",
            "connection": "pymongo",
            "kwargs": {
                "port": 27017,
                "password": "",
                "username": "",
                "host": "localhost"
            },
            "name": "documents"
        },
        "metadata": {
            "cls": "mongodb",
            "connection": "pymongo",
            "kwargs": {
                "port": 27017,
                "password": "",
                "username": "",
                "host": "localhost"
            },
            "name": "documents"
        }
    },
    "distributed": false,
    "cdc": false,
    "server": {
        "web_server": {
            "host": "127.0.0.1",
            "port": 3223,
            "protocol": "http"
        },
        "test_port": 32233
    },
    "vector_search": {
        "lancedb": {
            "uri": "./.lancedb"
        }
    }
}
```


superduperdb/base/config.py
"""
The classes in this file define the configuration variables for SuperDuperDB,
which means that this file gets imported before alost anything else, and
canot contain any other imports from this project.
"""

import typing as t
from enum import Enum

from pydantic import Field, root_validator

from .jsonable import Factory, JSONable

_BAD_KEY = '...bad.key...'
REST_API_VERSION = '0.1.0'


class HasPort(JSONable):
    """A base class for server connections"""

    password: str = ''
    port: int = 0
    username: str = ''


class HostPort(HasPort):
    """Configure server connections with a `host` property"""

    host: str = 'localhost'


class IpPort(HasPort):
    """Configure server connections with an `ip` property"""

    ip: str = 'localhost'


class Retry(JSONable):
    """Describes how to retry using the `tenacity` library"""

    stop_after_attempt: int = 2
    wait_max: float = 10.0
    wait_min: float = 4.0
    wait_multiplier: float = 1.0


class Api(JSONable):
    """A base class for API connections"""

    api_key: str = Field(default=_BAD_KEY, repr=False)


class Apis(JSONable):
    """A container for API connections"""

    providers: t.Dict[str, Api] = Factory(dict)
    retry: Retry = Factory(Retry)


class Dask(IpPort):
    """Describes a connection to Dask"""

    deserializers: t.List[str] = Factory(list)
    port: int = 8786
    serializers: t.List[str] = Factory(list)
    local: bool = True


class Deployment(JSONable):
    """(unused)"""

    database: str = ''
    model: str = ''


class LogLevel(str, Enum):
    """Enumerate log severity level"""

    DEBUG = 'DEBUG'
    INFO = 'INFO'
    WARN = 'WARN'


class LogType(str, Enum):
    """Enumerate the standard logs"""

    STDERR = 'STDERR'
    LOGGING = 'LOGGING'


class Logging(JSONable):
    """Describe how we are going to log. This isn't yet used."""

    level: LogLevel = LogLevel.INFO
    type: LogType = LogType.STDERR
    kwargs: dict = Factory(dict)


class ModelServer(HostPort):
    """Configure the model server's port"""

    host: str = '127.0.0.1'
    port: int = 5001


class MongoDB(HostPort):
    """Configure MongoDB's port and credentials"""

    host: str = 'localhost'
    port: int = 27017


class DataLayer(JSONable):
    """Configure which db or database is being used"""

    cls: str = 'mongodb'
    connection: str = 'pymongo'
    kwargs: t.Dict = Factory(lambda: MongoDB().dict())
    name: str = 'test_db'


class DataLayers(JSONable):
    """TBD"""

    artifact: DataLayer = Factory(lambda: DataLayer(name='_filesystem:test_db'))
    data_backend: DataLayer = Factory(DataLayer)
    metadata: DataLayer = Factory(DataLayer)


class Notebook(JSONable):
    """Configure the notebook server connection information"""

    ip: str = '0.0.0.0'
    password: str = ''
    port: int = 8888
    token: str = ''

    @root_validator(skip_on_failure=True)  # type: ignore[call-overload]
    def check_one_or_none(cls, v: t.Dict) -> t.Dict:
        if v['password'] and v['token']:
            raise ValueError('At most one of password and token may be set')
        return v


class Server(JSONable):
    """Configure the SuperDuperDB server connection information"""

    host: str = '127.0.0.1'
    port: int = 3223
    protocol: str = 'http'

    @property
    def uri(self) -> str:
        return f'{self.protocol}://{self.host}:{self.port}'


class LanceDB(JSONable):
    """Configure the Lance DB vector search connection information"""

    backfill_batch_size: int = 100
    lancedb: bool = True
    uri: str = './.lancedb'


class InMemory(JSONable):
    """Configure the in-memory vector search connection information"""

    backfill_batch_size: int = 100
    inmemory: bool = True


class VectorSearch(JSONable):
    """Configure the full vector search connection information"""

    host: str = 'localhost'
    password: str = Field(default='', repr=False)
    port: int = 19530
    type: t.Union[LanceDB, InMemory] = Field(default_factory=InMemory)
    username: str = Field(default='', repr=False)


class Config(JSONable):
    """The data class containing all configurable superduperdb values"""

    apis: Apis = Factory(Apis)
    cdc: bool = False
    dask: Dask = Factory(Dask)
    data_layers: DataLayers = Factory(DataLayers)
    distributed: bool = False
    logging: Logging = Factory(Logging)
    model_server: ModelServer = Factory(ModelServer)
    notebook: Notebook = Factory(Notebook)
    server: Server = Factory(Server)
    vector_search: VectorSearch = Factory(VectorSearch)
    verbose: bool = False

    class Config(JSONable.Config):
        protected_namespaces = ()
